### DISCUSSION TITLE
Address Visibility Across Control Plane & Data Plane

### DISCUSSION THREAD

**Discussion by simanadler** [fybrik/fybrik#890] [Developers]
The control plane and the data plane generate lots of information - status & virtual endpoints to users, errors, logs, events, metrics - each of which is meant for different users.  

We need to standardize the information that is generated and ensure:
* A standard format and mandatory content (timestamp, workload id, actor to which it's relevant ...)
* it is tagged appropriately so that we can ensure it can reach the right people
* it's level of priority is indicated. Ex: debugging info should be able to be turned off in production
* different levels/types of messages should have different retention periods
* messages that require internationalization (all user facing ones certainly) should be flagged and handled as such (i.e. no hard coded strings).  
* Unique ids to link information generated across components

The following describes approaches to handling such information, but not about how to generate it: https://docs.google.com/document/d/1gtKD6n_jWqDttiZw2A2BUHyNnAbaSSgq_z4Pif6mUQE/edit?usp=sharing

Related issues currently open:
#153 
#568 
#673 
#876 

Need to determine how to get the relevant messages to the relevant roles (user, IT admin, governance officer,...), how long to store them, etc.

@laurieaw63 Does your team have a standard approach to this?
@roee88 Are you aware of any standards or open source approaches for generating the messages?

**Comment by simanadler** [comment]
I think the first step is to ensure that the code generates error and log information in a manner that is as consistent and will enable cross correlation of logged issues across components.  Reviewing the current code I note the following:

**Loggers used:**
* manager https://github.com/go-logr/logr
    * Name of component (ex: blueprint) and namespace name are included
* Razee interface https://github.com/go-logr/logr 
* Connectors standard golang log library
* (Research) Arrow flight uses python logger, with the following formatting
loglevel=getattr(logging, args.loglevel.upper(), logging.WARNING)
logging.basicConfig(format='%(asctime)s - %(levelname)s - %(message)s', datefmt='%d/%m/%Y %H:%M:%S', level=loglevel)
 
**Logging issues**
1. There should be a unique IDs for the workload and dataset where relevant, so cross correlation across logs of different components can be done in the future.
    a. If logging something associated with the control plane deployment, that should be clearly indicated … including a unique id for the deployment 
2. Should include a UTC timestamp 
3. Need to support logging levels: error, info, debugging
Currently we are putting all of the log messages at info level 0 currently in the manager.  I.e. There is no differentiation between errors, info and debugging
 
**User Errors/Status Messages:** 
* In future will need to support internationalization, i.e. put messages that will be seen by end  users in string file that can be translated.

@roee88 @ronenkat @laurieaw63 @revit13 @cdoron @shlomitk1 your thoughts?

**Reply by roee88** [reply]
1. Adding to your message: Currently a lot of the log records are not self contained (logging is used as if it's printing one line after the other to stdout).
2. Logging is not intended for end users, so I don't see any need for internationalization. 
3. What we need is a concrete suggestion for how a log record looks like in a way that would be useful to easily filter records for debugging issues. You already mentioned filtering by fybrik application, asset, time. @shlomitk1 @revit13 when you debug an issue what information is usually useful for you?

**Reply by simanadler** [reply]
>     1. Adding to your message: Currently a lot of the log records are not self contained (logging is used as if it's printing one line after the other to stdout).
> 
>     2. Logging is not intended for end users, so I don't see any need for internationalization.
True.  I was referring to statuses and errors sent to the user.
> 
>     3. What we need is a concrete suggestion for how a log record looks like in a way that would be useful to easily filter records for debugging issues. You already mentioned filtering by fybrik application, asset, time. @shlomitk1 @revit13 when you debug an issue what information is usually useful for you?

I agree.  That's what this discussion will hopefully lead us towards.  I would add to your list: 
* control plane instance (for the future when more than one is supported)
* log event level (error, info, debug)

**Reply by roytman** [reply]
I'd suggest an additional level `audit`.  it can be targeted to the security/compliance officer, and probably retained.

**Comment by shlomitk1** [comment]
@roee88 Most of the errors go directly to the error report (inside fybrikapplication), thus printing errors is redundant. It is useful to see responses from connectors (enforcement actions, dataset geography, etc.), what clusters are deployed, what modules are available. When module selection fails, the decision process is important (e.g. found a read module that supports interface X, copy is required because interfaces do not match, ...). Tracing uuid is really important because otherwise it's impossible to understand to what application the log message refers.

**Reply by simanadler** [reply]
@shlomitk1 I think the logs will be used to understand in retrospect what happened with a given workload, so I think that error information is important to include as well.  Logs aren't only for debugging.  Right?

**Reply by roee88** [reply]
Thanks, that's really useful. Note that errors in the logs are not redundant, it's what you have to see what actually happened. You won't necessarily have access to the FybrikApplication and also in some cases the status already changed and you need to understand what happened at a specific time. Having all information in the logs is the right approach.

**Reply by shlomitk1** [reply]
@simanadler What I meant is that personally, I am not using error information in logs for debugging, but I agree that all information should be included.

**Reply by smshao** [reply]
From past experience, we have searched for errors as a starting point for problem determination. They may be useful in helping us find the potential problem areas quickly, and then we'll use the traces for problem determination.

**Comment by simanadler** [comment]
See [this document](https://docs.google.com/document/d/1x3Jnp3CIN6YhJKtWM1uIimaw9-q6Hb9M7LqkazmBRBo/edit?usp=sharing) for a summary of requirements and a high level design.  @roee88 @ronenkat @elsalant @laurieaw63

**Reply by smshao** [reply]
Some of the discussion in this issue brought up some interesting points in regards to logging.

First of all, there may be 3 types of logs and target audiences. The first is for debugging and those logs are meant for the developers. This type of logging is usually not translated and is logged in the language of developer. The second type is operational meant for system administrators. These log entries let the administrators know about events in the system to allow them to take corrective actions. These log messages are generally logged in the language of the user, therefore, translated. The third type is audit meant for security and data administrators. These log entries allow the data administrator to determine if the policies they authored has the correct effect, and security administrators to determine which user performed which actions. These log entries are usually logged in the user's language.

**Reply by simanadler** [reply]
Thanks @smshao !  We were indeed wondering whether we needed to support internationalization of log messages.  My initial thought was yes, but others pointed out that other infrastructure level component such as kubernetes do not translate log messages.  Does RedHat open shift translate log messages? @roytman @ronenkat

**Reply by roytman** [reply]
I remember that in WebSphere all error messages had ID and were translated to different native languages. In the case of OpenShift, I did not see this pattern. All messages are in English. 
Checked with Eran Rainshtein, his team works with the OpenShift logging team 
Q: Do you know if RedHat plans to support NLP (native language processing) of log messages ?
A: They do not and probably will never be in this space, they keep that for 3rd parties

**Comment by simanadler** [comment]
I chatted with @eranra who is working with OpenShift on logging.  (Thanks for the suggestion  @Alexey) The feedback I got is as follows: (1) JSON is most common today, so this is almost trivial to suggest that as the high level format. (2) As for localisation, I am almost sure that this is not needed, those infrastructure log lines are for admins and there is an assumption that they work in English. I would not suggest to spend to time on that, unless those are very “customer oriented” logs lines (3) as for the list of fields, you might want to consolidate several of the mandatory fields such as component, function and maybe also cluster into one “path” like context field.

**Comment by simanadler** [comment]
As we move to a fybrik logging library that can be used by both the control plane and by data plane components, I’m thinking it should be in a dedicated fybrik/logging repository rather than in the main fybrik repository under pkgs.  Do you agree?  @Alexey @Ronen Kat@eres @shlomitk @rovallam @Rachit Arora @cdoron

**Reply by ronenkat** [reply]
@simanadler I would suggest starting in the Fybrik repo as the main client of logging will be the control plane. It is also a manner of velocity and managing resources.
A separate repo implies, CI workflow, release cycle, packaging, etc --- a lot of overhead, and as we will use existing logging toolkits, we should only have wrapper code - should be a lot of overhead.

More so, the data plane is not language specific, we have Python, Go and Java --- will you develop libraries for all of these? so for data plane I would define the logging spec, and as those specs are being implemented, review your suggestion again.

**Reply by simanadler** [reply]
@ronenkat Yes, the intent is to develop for Python, Go and Java.  We will start with go.  @roytman Are you in agreement with @ronenkat about putting the logging in the main Fybrik repo rather than in a stand-alone repo?

**Reply by roytman** [reply]
Hi @simanadler , yes let's start from the main fybrik repository. If we see that it is required by other repos, we always can separate it. 
BTW, do we know any other components that can be used by different repos ?

**Reply by roytman** [reply]
A question, should fybrik provide a mechanism to separate log access? I mean that a Data Scientist will not be able to see auditing logs targeted to a security/complience officer.

**Reply by shlomitk1** [reply]
> A question, should fybrik provide a mechanism to separate log access? I mean that a Data Scientist will not be able to see auditing logs targeted to a security/complience officer.

Would a Data Scientist see any logs? Not sure he/she gets a permission to the control-plane namespace. I'd assume the logs are aimed for IT admin, developers, security officers, etc. Users get the information propagated to the application status, possibly via some UI.

**Reply by simanadler** [reply]
> Hi @simanadler , yes let's start from the main fybrik repository. If we see that it is required by other repos, we always can separate it. BTW, do we know any other components that can be used by different repos ?

We already have components outside of the main fybrik repo that should use this: data-movement-operator, arrow-flight, and the example code for developing modules.

**Reply by simanadler** [reply]
> > A question, should fybrik provide a mechanism to separate log access? I mean that a Data Scientist will not be able to see auditing logs targeted to a security/complience officer.
> 
> Would a Data Scientist see any logs? Not sure he/she gets a permission to the control-plane namespace. I'd assume the logs are aimed for IT admin, developers, security officers, etc. Users get the information propagated to the application status, possibly via some UI.

Different actors will need to see different information.  For some actors the source of the information they see may come from logs and for others (such as users) it may come from fybrikapplication.status.  Please see the description in the [overview document ](https://docs.google.com/document/d/1x3Jnp3CIN6YhJKtWM1uIimaw9-q6Hb9M7LqkazmBRBo/edit#heading=h.z3iyvac259h6).  In any case, it's not the role of fybrik to determine how to get the necessary information to the different actors, but rather to ensure that the log entries contain the information needed so that it can be easily filtered by whatever log aggregator tool is chosen.

For example, there is a project under way that needs to take fybrik user level log entries and copy them to a blockchain that is being used for auditing purposes.  This approach is preferable to them since they do not want to access kubernetes - i.e. fybrikapplication.status.  Instead they want to create a fluentd filter to get the relevant info from the logs.  In this manner they can get user level log entries using one filter rather than monitoring the fybrikapplication.status and the arrow-flight status.

**Reply by shlomitk1** [reply]
@simanadler I think the term user is misleading. What is copied to a blockchain or archived is not a set of user level log entries. This is the information required for audit purposes. For example, types of actions and algorithms that were used, additional storage that was created, deleted, etc. - these are not details that interest a Data Scientist.
